File: screens/profile_screen.py
import flet as ft

from services.db_service import get_preferences, update_preference
from theme import colors


def build_profile_screen(page: ft.Page) -> ft.Container:
    prefs = get_preferences()

    goal_text = ft.Text(f"{prefs['daily_goal']} min", size=14, weight=ft.FontWeight.W_500, color=colors.TEXT_PRIMARY)

    def change_goal(delta: int):
        new_value = max(5, min(60, prefs["daily_goal"] + delta))
        prefs["daily_goal"] = new_value
        goal_text.value = f"{new_value} min"
        update_preference("daily_goal", new_value)
        page.update()

    def build_toggle_row(icon_name: str, title: str, subtitle: str, key: str) -> ft.Container:
        def on_toggle(e):
            prefs[key] = e.control.value
            update_preference(key, e.control.value)

        return ft.Container(
            padding=ft.Padding.symmetric(vertical=10),
            content=ft.Row(
                vertical_alignment=ft.CrossAxisAlignment.CENTER,
                controls=[
                    ft.Icon(getattr(ft.Icons, icon_name.upper()), size=20, color=colors.BLUE_ACCENT),
                    ft.Column(
                        expand=True,
                        spacing=1,
                        controls=[
                            ft.Text(title, size=14, weight=ft.FontWeight.W_500, color=colors.TEXT_PRIMARY),
                            ft.Text(subtitle, size=11, color=colors.TEXT_SECONDARY),
                        ],
                    ),
                    ft.Switch(value=prefs[key], active_color=colors.BLUE_ACCENT, on_change=on_toggle),
                ],
            ),
        )

    def on_logout(e):
        print("Cerrar sesion presionado")

    return ft.Container(
        expand=True,
        bgcolor=colors.BG_PAGE,
        padding=ft.Padding.only(left=20, right=20, top=20),
        content=ft.Column(
            expand=True,
            spacing=16,
            scroll=ft.ScrollMode.AUTO,
            controls=[
                ft.Column(
                    spacing=2,
                    controls=[
                        ft.Text("TU CUENTA", size=12, weight=ft.FontWeight.W_500, color=colors.BLUE_ACCENT),
                        ft.Text("Perfil", size=24, weight=ft.FontWeight.W_500, color=colors.TEXT_PRIMARY),
                    ],
                ),
                ft.Container(
                    bgcolor=colors.BG_CARD,
                    border_radius=16,
                    padding=16,
                    content=ft.Row(
                        spacing=12,
                        vertical_alignment=ft.CrossAxisAlignment.CENTER,
                        controls=[
                            ft.Container(
                                width=48,
                                height=48,
                                border_radius=24,
                                bgcolor=colors.BLUE_ACCENT,
                                alignment=ft.Alignment.CENTER,
                                content=ft.Text("J", size=18, weight=ft.FontWeight.W_500, color=colors.BG_PAGE),
                            ),
                            ft.Column(
                                spacing=1,
                                controls=[
                                    ft.Text("Jair", size=15, weight=ft.FontWeight.W_500, color=colors.TEXT_PRIMARY),
                                    ft.Text("Miembro desde agosto 2026", size=11, color=colors.TEXT_SECONDARY),
                                ],
                            ),
                        ],
                    ),
                ),
                ft.Container(
                    bgcolor=colors.BG_CARD,
                    border_radius=16,
                    padding=16,
                    content=ft.Column(
                        spacing=4,
                        controls=[
                            ft.Row(
                                alignment=ft.MainAxisAlignment.SPACE_BETWEEN,
                                vertical_alignment=ft.CrossAxisAlignment.CENTER,
                                controls=[
                                    ft.Column(
                                        spacing=1,
                                        controls=[
                                            ft.Text("Meta diaria", size=14, weight=ft.FontWeight.W_500, color=colors.TEXT_PRIMARY),
                                            ft.Text("Minutos de practica", size=11, color=colors.TEXT_SECONDARY),
                                        ],
                                    ),
                                    ft.Row(
                                        spacing=8,
                                        controls=[
                                            ft.IconButton(
                                                icon=ft.Icons.REMOVE,
                                                icon_size=16,
                                                icon_color=colors.TEXT_SECONDARY,
                                                bgcolor=colors.BG_CARD_ICON,
                                                on_click=lambda e: change_goal(-5),
                                            ),
                                            goal_text,
                                            ft.IconButton(
                                                icon=ft.Icons.ADD,
                                                icon_size=16,
                                                icon_color=colors.TEXT_SECONDARY,
                                                bgcolor=colors.BG_CARD_ICON,
                                                on_click=lambda e: change_goal(5),
                                            ),
                                        ],
                                    ),
                                ],
                            ),
                            ft.Divider(color=colors.BG_CARD_ICON, height=1),
                            build_toggle_row("notifications", "Recordatorios diarios", "8:00 p. m.", "reminders"),
                            build_toggle_row("music_note", "Sonidos suaves", "Guia auditiva en ejercicios", "sounds"),
                            build_toggle_row("chat_bubble_outline", "Compartir con terapeuta", "Enviar resumen semanal", "share_therapist"),
                        ],
                    ),
                ),
                ft.Container(
                    bgcolor=colors.BG_CARD,
                    border_radius=12,
                    padding=14,
                    alignment=ft.Alignment.CENTER,
                    on_click=on_logout,
                    content=ft.Text("Cerrar sesion", size=14, weight=ft.FontWeight.W_500, color=colors.TEXT_PRIMARY),
                ),
            ],
        ),
    )

File: app/src/main/java/com/example/serena/ui/screens/ProfileScreen.kt
package com.example.serena.ui.screens

import android.Manifest
import android.content.pm.PackageManager
import android.os.Build
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Chat
import androidx.compose.material.icons.filled.MusicNote
import androidx.compose.material.icons.filled.Notifications
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.Switch
import androidx.compose.material3.SwitchDefaults
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.content.ContextCompat
import com.example.serena.ReminderScheduler
import com.example.serena.data.AppDatabase
import com.example.serena.data.UserPreferencesEntity
import com.example.serena.ui.theme.*
import kotlinx.coroutines.launch

@Composable
fun ProfileScreen() {
    val context = LocalContext.current
    val db = remember { AppDatabase.getInstance(context) }
    val scope = rememberCoroutineScope()

    var dailyGoal by remember { mutableStateOf(15) }
    var remindersEnabled by remember { mutableStateOf(true) }
    var soundsEnabled by remember { mutableStateOf(true) }
    var shareTherapist by remember { mutableStateOf(false) }
    var loaded by remember { mutableStateOf(false) }

    val notificationPermissionLauncher = rememberLauncherForActivityResult(
        ActivityResultContracts.RequestPermission()
    ) { granted ->
        if (granted) {
            scope.launch { ReminderScheduler.scheduleNext(context, hour = 20, minute = 0) }
        }
    }

    LaunchedEffect(Unit) {
        val saved = db.userPreferencesDao().get()
        if (saved != null) {
            dailyGoal = saved.dailyGoal
            remindersEnabled = saved.remindersEnabled
            soundsEnabled = saved.soundsEnabled
            shareTherapist = saved.shareTherapist
        }
        loaded = true
    }

    fun persist() {
        if (!loaded) return
        scope.launch {
            db.userPreferencesDao().save(
                UserPreferencesEntity(
                    dailyGoal = dailyGoal,
                    remindersEnabled = remindersEnabled,
                    soundsEnabled = soundsEnabled,
                    shareTherapist = shareTherapist
                )
            )
        }
    }

    fun onRemindersToggle(enabled: Boolean) {
        remindersEnabled = enabled
        persist()
        if (enabled) {
            val hasPermission = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                ContextCompat.checkSelfPermission(context, Manifest.permission.POST_NOTIFICATIONS) == PackageManager.PERMISSION_GRANTED
            } else {
                true
            }
            if (hasPermission) {
                scope.launch { ReminderScheduler.scheduleNext(context, hour = 20, minute = 0) }
            } else {
                notificationPermissionLauncher.launch(Manifest.permission.POST_NOTIFICATIONS)
            }
        } else {
            ReminderScheduler.cancel(context)
        }
    }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(BgPage)
            .padding(horizontal = 20.dp, vertical = 16.dp)
    ) {
        Text("TU CUENTA", color = BlueAccent, fontSize = 11.sp)
        Text("Perfil", color = TextPrimary, fontSize = 22.sp, fontWeight = FontWeight.Bold)
        Spacer(modifier = Modifier.height(16.dp))

        Row(
            modifier = Modifier
                .fillMaxWidth()
                .background(BgCard, shape = RoundedCornerShape(16.dp))
                .padding(16.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Box(
                modifier = Modifier
                    .size(52.dp)
                    .background(BlueAccent, shape = CircleShape),
                contentAlignment = Alignment.Center
            ) {
                Text("J", color = BgPage, fontSize = 19.sp, fontWeight = FontWeight.Bold)
            }
            Spacer(modifier = Modifier.width(12.dp))
            Column {
                Text("Jair", color = TextPrimary, fontSize = 15.sp, fontWeight = FontWeight.Bold)
                Text("Miembro desde agosto 2026", color = TextSecondary, fontSize = 11.sp)
            }
        }
        Spacer(modifier = Modifier.height(12.dp))

        Column(
            modifier = Modifier
                .fillMaxWidth()
                .background(BgCard, shape = RoundedCornerShape(16.dp))
                .padding(16.dp)
        ) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                verticalAlignment = Alignment.CenterVertically
            ) {
                Column(modifier = Modifier.weight(1f)) {
                    Text("Meta diaria", color = TextPrimary, fontSize = 15.sp, fontWeight = FontWeight.Bold)
                    Text("Minutos de practica", color = TextSecondary, fontSize = 11.sp)
                }
                IconButton(onClick = {
                    dailyGoal = (dailyGoal - 5).coerceAtLeast(5)
                    persist()
                }) {
                    Text("-", color = TextSecondary, fontSize = 18.sp)
                }
                Text("$dailyGoal min", color = TextPrimary, fontSize = 14.sp, fontWeight = FontWeight.Bold)
                IconButton(onClick = {
                    dailyGoal = (dailyGoal + 5).coerceAtMost(60)
                    persist()
                }) {
                    Text("+", color = TextSecondary, fontSize = 18.sp)
                }
            }
            Spacer(modifier = Modifier.height(8.dp))
            ToggleRow(Icons.Filled.Notifications, "Recordatorios diarios", "8:00 p. m.", remindersEnabled) { onRemindersToggle(it) }
            Spacer(modifier = Modifier.height(14.dp))
            ToggleRow(Icons.Filled.MusicNote, "Sonidos suaves", "Guia auditiva en ejercicios", soundsEnabled) {
                soundsEnabled = it
                persist()
            }
            Spacer(modifier = Modifier.height(14.dp))
            ToggleRow(Icons.Filled.Chat, "Compartir con terapeuta", "Enviar resumen semanal", shareTherapist) {
                shareTherapist = it
                persist()
            }
        }
        Spacer(modifier = Modifier.height(12.dp))

        Box(
            modifier = Modifier
                .fillMaxWidth()
                .background(BgPage, shape = RoundedCornerShape(14.dp))
                .clickable { }
                .padding(vertical = 14.dp),
            contentAlignment = Alignment.Center
        ) {
            Text("Cerrar sesion", color = TextPrimary, fontSize = 14.sp, fontWeight = FontWeight.Bold)
        }
    }
}

@Composable
private fun ToggleRow(
    icon: androidx.compose.ui.graphics.vector.ImageVector,
    title: String,
    subtitle: String,
    checked: Boolean,
    onCheckedChange: (Boolean) -> Unit
) {
    Row(verticalAlignment = Alignment.CenterVertically) {
        Icon(icon, contentDescription = null, tint = BlueAccent, modifier = Modifier.size(22.dp))
        Spacer(modifier = Modifier.width(10.dp))
        Column(modifier = Modifier.weight(1f)) {
            Text(title, color = TextPrimary, fontSize = 15.sp, fontWeight = FontWeight.Bold)
            Text(subtitle, color = TextSecondary, fontSize = 11.sp)
        }
        Switch(
            checked = checked,
            onCheckedChange = onCheckedChange,
            colors = SwitchDefaults.colors(
                checkedThumbColor = BgPage,
                checkedTrackColor = BlueAccent,
                uncheckedThumbColor = TextSecondary,
                uncheckedTrackColor = BgCardIcon
            )
        )
    }
}